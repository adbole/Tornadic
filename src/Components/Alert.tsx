import React from 'react';
import { AlertType, WeatherData } from '../ts/WeatherData';
import { NWSAlert, useWeather } from './Contexes/WeatherContext';
import { Modal, useModal } from './Contexes/ModalContext';
import { Widget, WidgetSize } from './SimpleComponents';

/**
 * Takes an ISO8601 string and converts it to a nice readable string in the format weekday, month, day, hour:minute AM/PM timezone
 * @param iso The ISO8601 string to convert
 * @returns The now nice readable string obtained from conversion
 */
function GetTimeString(iso: string) {
    return new Date(iso).toLocaleTimeString("en-us", {weekday:"short", month:"short", day:"numeric", hour12:true, hour:"numeric", minute:"numeric", timeZoneName:"short"});
}

function ToAlertCSS(alert: AlertType) {
    switch(alert) {
        case AlertType.WARNING:
            return "alert-warning";
        case AlertType.WATCH:
            return "alert-watch";
        case AlertType.ADVISORY:
            return "alert-advisory";
        default:
            return "alert-none";
    }
}

const AlertDisplay = ({alertData, className, onClick} : {alertData: NWSAlert, className: string, onClick: React.MouseEventHandler<HTMLDivElement>}) => (
    <Widget size={WidgetSize.WIDE} id="alert" className={className} onClick={onClick}>
        <h2>{alertData.properties.event}</h2>
        <p>{alertData.properties.event} until {GetTimeString(alertData.properties.ends ?? alertData.properties.expires)}</p>
    </Widget>
);

const Alert = () => {
    const alertData = useWeather().alerts;
    const modals = useModal();

    //If no alerts are active then don't display this component.
    if(!alertData.length) return <></>;

    //Determine which alert should be shown. Warnings get highest priority with the lastest being first. 
    let alertToShow = alertData[0];
    let highestAlertType = WeatherData.GetAlertType(alertData[0]);
    let newestAge = new Date(alertData[0].properties.sent);

    for(let i = 1; i < alertData.length; ++i) {
        const ithType = WeatherData.GetAlertType(alertData[i]);
        const sentDate = new Date(alertData[i].properties.sent);

        //Replace the alertToShow if its severity is lower
        if(highestAlertType < ithType) {
            alertToShow = alertData[i];
            highestAlertType = ithType;
            newestAge = sentDate;
        }
        //When two alerts are the same severity, we want the youngest one
        else if(highestAlertType === ithType && sentDate > newestAge) {
            alertToShow = alertData[i];
            highestAlertType = ithType;
            newestAge = sentDate;
        }
    }

    const onClickHandler = () => modals.showModal(alertData.length > 1 ? <AlertSelectionModal alertData={alertData}/> : <AlertModal alert={alertData[0]}/>);

    return <AlertDisplay alertData={alertToShow} className={ToAlertCSS(highestAlertType)} onClick={onClickHandler}/>;
};

export default Alert;

const AlertSelectionModal = ({alertData}: {alertData: NWSAlert[]}) => {
    const modals = useModal();

    //Memoize the component to prevent unnecessary operations
    const memoizedComponent = React.useMemo(() => (
        <Modal modalTitle={alertData.length + " Weather Alerts"} modalTitleClass={"alert-none"} id="alert-list-modal">
            {
                alertData.map((alert, key) => {
                    //Inclues override to onClose to reopen the alert selection modal
                    const onClickHandler = () => modals.showModal(<AlertModal alert={alert} onClose={() => modals.showModal(memoizedComponent)}/>);

                    return <AlertDisplay key={key} alertData={alert} className={ToAlertCSS(WeatherData.GetAlertType(alert))} onClick={onClickHandler}/>;
                })
            }
        </Modal>
    ), [alertData]);
    

    return memoizedComponent;
};

/**
 * Display the given NWSAlert as a modal to get more in depth information from.
 */
export const AlertModal = (props: {alert: NWSAlert} & React.DialogHTMLAttributes<HTMLDialogElement>) => {
    const {alert, ...excess} = props;

    const alertData = alert.properties;
    const alertType = ToAlertCSS(WeatherData.GetAlertType(alert));
    
    return (
        <Modal modalTitle={alertData.event} modalTitleClass={alertType} id="alert-modal" {...excess}>
            <p><em>Issuing Office:</em> {alertData.senderName}</p>
            <p><em>Issued:</em> {GetTimeString(alertData.sent)}</p>
            <p><em>Effective:</em> {GetTimeString(alertData.effective)}</p>
            <p><em>Ends:</em> {GetTimeString(alertData.ends ?? alertData.expires)}</p>
            <hr/>
            {
                alertData.instruction && 
                <>
                    <h3>Instructions</h3>
                    <p>{alertData.instruction}</p>
                    <hr/>
                </>
            }
            {
                alertData.description.split("*").map((string, index) => {
                    if(!string.length) return <React.Fragment key={index}/>;
        
                    return <p key={index}>* {string}</p>;
                })    
            }
            <hr/>
            <h3>Affected Areas</h3>
            <p>{alertData.areaDesc}</p>
        </Modal>
    );
};