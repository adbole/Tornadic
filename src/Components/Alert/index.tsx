import React from "react";
import booleanPointInPolygon from "@turf/boolean-point-in-polygon";

import { useBooleanState, usePermission } from "Hooks";

import { useWeather } from "Contexts/WeatherContext";

import AlertModal from "Components/Modals/Alert";

import type NWSAlert from "ts/NWSAlert";

import AlertWidget, { AlertInformation, ExcessAlerts } from "./style";


const getPriorityAlert = (alerts: NWSAlert[]) =>
    alerts.reduce(
        (highest, next) => (next.priority < highest.priority ? next : highest),
        alerts[0]
    );

function Alert({ noNotify = false }) {
    const { alerts: unfilteredAlerts, point } = useWeather();
    const notiPermission = usePermission("notifications")

    const [modalOpen, showModal, hideModal] = useBooleanState(false);

    const alerts = unfilteredAlerts.filter(alert => {
        //Point and NWSAlert include the necessary GeoJSON properties for this to work so conversion to any is okay
        const inBounds = alert.hasCoords() && booleanPointInPolygon(point as any, alert as any)
            
        return alert.get("affectedZones").includes(point.properties.forecastZone) || inBounds
    });

    const previousAlertTime = React.useRef(new Date(alerts[0]?.get("sent") ?? 0));

    React.useEffect(() => {
        if (noNotify || alerts.length === 0 || notiPermission !== "granted") return;

        const newAlerts = alerts.filter((alert) => new Date(alert.get("sent")) > previousAlertTime.current);

        if (newAlerts.length > 0) {
            const alert = getPriorityAlert(newAlerts);

            const additionalAlerts = newAlerts.length > 1 ? `\n+${newAlerts.length - 1} more alert(s)` : "";

            new Notification(`${alert.get("event")}`, {
                body: `Issued: ${alert.get("sent")}\nExpires: ${alert.get("expires") ?? alert.get("ends")}${additionalAlerts}`,
            });

            previousAlertTime.current = new Date(newAlerts[0].get("sent"));
        }

    //This effect should only run if alerts changes, noNotify and notiPermission are excluded as changing them could 
    //cause unnecessary notifications to be sent annoying the user.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [alerts]);

    if (!alerts.length) return null;

    const alert = getPriorityAlert(alerts);

    return (
        <>
            <AlertWidget
                type={alert.getAlertCSS()}
                isTemplate
                size="widget-wide"
                onClick={showModal}
            >
                <AlertInformation>
                    <h1>{alert.get("event")}</h1>
                    <p>
                        <em>Issued:</em> {alert.get("sent")}
                    </p>
                    <p>
                        <em>Until:</em> {alert.get("ends") ?? alert.get("expires")}
                    </p>
                </AlertInformation>

                {alerts.length > 1 && (
                    <ExcessAlerts>+{alerts.length - 1} more alert(s)</ExcessAlerts>
                )}
            </AlertWidget>
            <AlertModal alerts={alerts} isOpen={modalOpen} onClose={hideModal} />
        </>
    );
}

export default Alert;
