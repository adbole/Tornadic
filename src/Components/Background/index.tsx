import React from "react";
import { Global } from "@emotion/react";
import styled from "@emotion/styled";
import { Sky } from "@react-three/drei";
import { Canvas } from "@react-three/fiber";

import { useReadLocalStorage } from "Hooks";

import { useWeather } from "Contexts/WeatherContext";

import { Clouds, Gradient, RainSnow, Stars, Thunder } from "./__internal__";


const StyledCanvas = styled(Canvas)({
    top: 0,
    left: 0,
    zIndex: -1,
});

function HighConstrast({ rayleigh }: { rayleigh: number }) {
    const settings = useReadLocalStorage("userSettings");

    if (rayleigh !== 1 || !settings?.highContrastForLive) return null;

    return <Global styles={{ "#root": { color: "#3f3f3f" } }} />;
}

export default function Background() {
    const { weather } = useWeather();
    const settings = useReadLocalStorage("userSettings");

    const { isDay, condition } = React.useMemo(
        () => ({
            isDay: weather.isDay(),
            condition: weather.getNow().conditionInfo.type,
        }),
        [weather]
    );

    const [rayleigh, elevation, exposure] = React.useMemo(() => {
        if (isDay) {
            switch (condition) {
                case "Clear":
                case "Mostly Clear":
                    return [1, 0.75, 0.25];
                default:
                    return [5, 0.75, 0.1];
            }
        }

        return [0.01, 0.75, 0.2];
    }, [isDay, condition]);

    return (
        <>
            <HighConstrast rayleigh={rayleigh} />
            <Gradient isDay={isDay} condition={condition} />
            
            {settings && !settings.preferGradient && (
                <StyledCanvas
                    style={{ position: "fixed" }}
                    gl={{ toneMappingExposure: exposure }}
                >
                    <Sky
                        rayleigh={rayleigh}
                        inclination={elevation}
                        mieCoefficient={0.005}
                        mieDirectionalG={0.99}
                        turbidity={1}
                        azimuth={0.25}
                    />
                    <Clouds isDay={isDay} condition={condition} />
                    <RainSnow condition={condition} />
                    <Thunder condition={condition} />
                    <Stars isDay={isDay} condition={condition} />
                </StyledCanvas>
            )}
        </>
    );
}
