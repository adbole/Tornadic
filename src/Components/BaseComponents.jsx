import PropTypes from 'prop-types';

export const Widget = (props) => {
    const {className, children, large, ...excess} = props;

    let defaultClass = "widget "
    if(large) defaultClass += "widget-large "

    return (
        <div className={defaultClass + className} {...excess} >
            {children}
        </div>
    )
}

Widget.defaultProps = { className: '' }
Widget.propTypes = { large: PropTypes.bool }

export const OrderedFlexList = (props) => (
    <ol className={"flex-list flex-list-" + props.type}>
        {props.children}
    </ol>
)

OrderedFlexList.propTypes = {
    type: PropTypes.oneOf(['row', 'column']).isRequired
}

export const BasicInfoView = (props) => (
    <Widget className="basic-info">
        {props.icon}
        <p>{props.title}</p>
        <p>{props.value}</p>
    </Widget>
)

BasicInfoView.propTypes = {
    icon: PropTypes.element.isRequired,
    title: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired
}