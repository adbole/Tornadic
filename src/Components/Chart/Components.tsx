import { AreaChart, BarChart, LineChart, TooltipProps } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";
import { NameType } from "recharts/types/component/DefaultTooltipContent";

import { UV_MAX_VALUES } from "ts/Constants";

import { ChartViews, DataPoint } from ".";

function getUnit(data: DataPoint, defaultUnit: React.ReactNode) {
    const { property, primaryKey } = data;

    if(property === ChartViews.UV_Index) {
        if (primaryKey <= UV_MAX_VALUES.LOW) return " Low"; 
        else if (primaryKey <= UV_MAX_VALUES.MODERATE) return " Moderate";
        else if (primaryKey <= UV_MAX_VALUES.HIGH) return " High";
        else if (primaryKey <= UV_MAX_VALUES.VERY_HIGH) return " Very High";
        else return " Extreme"; 
    }

    return defaultUnit;
}
  
//Customized tooltip to display primary and secondary data with labels if applicable
export const CustomTooltip = ({ active, payload }: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const fixDecimal = (property: ChartViews, value: number): string | number => 
        (property === ChartViews.Precipitation || property === ChartViews.Pressure) ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    let unit = getUnit(data, payload[0].unit);
    const label = (data.property === ChartViews.Temperature && "Feels: ") || (data.property === ChartViews.Windspeed && "Gust: ");


    return (
        <div className="chart-tooltip">
            <h1>{fixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{fixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

type ChartDisplayProps = {
    children: React.ReactNode,
    property: ChartViews
} & CategoricalChartProps

//Provides the proper chart to display based on the HourlyProperty passed
export const ChartDisplay = (props: ChartDisplayProps) => {
    const { children, property, ...excess } = props;

    let ChosenChart = BarChart;

    switch(property) {
        case ChartViews.Precipitation:
            ChosenChart = BarChart;
            break;
        case ChartViews.Temperature:
        case ChartViews.UV_Index:
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return (
        <ChosenChart {...excess}>
            {children}
        </ChosenChart>
    );
};