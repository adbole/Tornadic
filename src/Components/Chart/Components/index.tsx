export { default as Area } from "./Area"
export { default as Axes } from "./Axes"
export { default as Bar } from "./Bar"
export { default as Line } from "./Line"
export { default as NowReference } from "./NowReference"
export { default as Tooltip, useTooltip } from "./Tooltip"