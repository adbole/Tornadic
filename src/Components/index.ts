export { default as Alert } from "./Alert";
export { default as Daily } from "./Daily";
export { default as HazardLevel } from "./HazardLevel";
export { default as Hourly } from "./Hourly";
export * from "./Input";
export { default as MessageScreen } from "./MessageScreen";
export { default as Chart } from "./Modals/Chart";
export { default as Modal } from "./Modals/Modal";
export * from "./Modals/Modal";
export { default as Now } from "./Now";
export { default as Pressure } from "./Pressure";
export { default as Radar } from "./Radar";
export { default as SimpleInfoWidget } from "./SimpleInfoWidget";
export { default as SkeletonWidget } from "./SkeletonWidget";
export { default as SunTime } from "./SunTime";
export { default as Widget } from "./Widget";
export { default as Wind } from "./Wind";
