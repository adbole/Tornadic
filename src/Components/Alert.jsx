import PropTypes from 'prop-types'
import { Widget } from "./BaseComponents";

const Alert = (props) => (
    <Widget className={"alert " + props.type}>
        <h2>{props.name}</h2>
        <p>{props.message}</p>
    </Widget>
)
Alert.propTypes = {
    type: PropTypes.oneOf(['warning', 'watch', 'advisory', 'special']).isRequired,
    name: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    moreExist: PropTypes.bool
}

export default Alert;