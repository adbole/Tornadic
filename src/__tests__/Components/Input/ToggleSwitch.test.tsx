import { act, render, screen } from "@testing-library/react";

import { ToggleSwitch } from "Components";


test("Matches snapshot", () => {
    const { container } = render(
        <ToggleSwitch 
            label="MyToggleSwitch"
            title="MyToggleSwitchTitle"
            onChange={() => undefined}
            defaultChecked={false}
        />
    )

    expect(container).toMatchSnapshot()
})

test("Is not checked when defaultChecked is false", () => {
    render(
        <ToggleSwitch 
            label="MyToggleSwitch"
            onChange={() => undefined}
            defaultChecked={false}
        />
    )

    expect(screen.getByLabelText("MyToggleSwitch")).not.toBeChecked()
})

test("Is checked by when defaultChecked is true", () => {
    render(
        <ToggleSwitch 
            label="MyToggleSwitch"
            onChange={() => undefined}
            defaultChecked={true}
        />
    )

    expect(screen.getByLabelText("MyToggleSwitch")).toBeChecked()
})

test("onChange is called when clicked", () => {
    const onChange = vi.fn()

    render(
        <ToggleSwitch 
            label="MyToggleSwitch"
            title="MyToggleSwitchTitle"
            onChange={onChange}
            defaultChecked={false}
        />
    )

    act(() => {
        screen.getByLabelText("MyToggleSwitch").click()
    })

    expect(onChange).toHaveBeenCalledOnce()
})