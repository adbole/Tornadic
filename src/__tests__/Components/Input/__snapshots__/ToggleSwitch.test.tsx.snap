// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Is checked by default 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: relative;
  width: 4rem;
  height: 2rem;
  border-radius: var(--border-radius);
  background-color: var(--background-layer);
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  outline: none;
  cursor: pointer;
}

.emotion-0::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 30%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border-radius: 50%;
  height: 1.25rem;
  width: 1.25rem;
  background-color: white;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.emotion-0:checked {
  background-color: var(--primary);
}

.emotion-0:checked::after {
  left: 70%;
}

<input
  checked=""
  class="emotion-0 emotion-1"
  type="checkbox"
/>
`;

exports[`Is checked by when defaultChecked is true 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: relative;
  width: 4rem;
  height: 2rem;
  border-radius: var(--border-radius);
  background-color: var(--background-layer);
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  outline: none;
  cursor: pointer;
}

.emotion-0::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 30%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border-radius: 50%;
  height: 1.25rem;
  width: 1.25rem;
  background-color: white;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.emotion-0:checked {
  background-color: var(--primary);
}

.emotion-0:checked::after {
  left: 70%;
}

<input
  checked=""
  class="emotion-0 emotion-1"
  type="checkbox"
/>
`;

exports[`Matches default checked snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  margin: 5px 0px;
}

.emotion-2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: relative;
  width: 4rem;
  height: 2rem;
  border-radius: var(--border-radius);
  background-color: var(--background-layer);
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  outline: none;
  cursor: pointer;
}

.emotion-2::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 30%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border-radius: 50%;
  height: 1.25rem;
  width: 1.25rem;
  background-color: white;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.emotion-2:checked {
  background-color: var(--primary);
}

.emotion-2:checked::after {
  left: 70%;
}

<div>
  <label
    class="emotion-0 emotion-1"
  >
    <input
      checked=""
      class="emotion-2 emotion-3"
      type="checkbox"
    />
    <span>
      MyToggleSwitch
    </span>
  </label>
</div>
`;

exports[`Matches default snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  margin: 5px 0px;
}

.emotion-2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: relative;
  width: 4rem;
  height: 2rem;
  border-radius: var(--border-radius);
  background-color: var(--background-layer);
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  outline: none;
  cursor: pointer;
}

.emotion-2::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 30%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border-radius: 50%;
  height: 1.25rem;
  width: 1.25rem;
  background-color: white;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.emotion-2:checked {
  background-color: var(--primary);
}

.emotion-2:checked::after {
  left: 70%;
}

<div>
  <label
    class="emotion-0 emotion-1"
    title="MyToggleSwitchTitle"
  >
    <input
      class="emotion-2 emotion-3"
      type="checkbox"
    />
    <span>
      MyToggleSwitch
    </span>
  </label>
</div>
`;

exports[`Matches snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  margin: 5px 0px;
}

.emotion-2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: relative;
  width: 4rem;
  height: 2rem;
  border-radius: var(--border-radius);
  background-color: var(--background-layer);
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  outline: none;
  cursor: pointer;
}

.emotion-2::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 30%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border-radius: 50%;
  height: 1.25rem;
  width: 1.25rem;
  background-color: white;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.emotion-2:checked {
  background-color: var(--primary);
}

.emotion-2:checked::after {
  left: 70%;
}

<div>
  <label
    class="emotion-0 emotion-1"
    title="MyToggleSwitchTitle"
  >
    <input
      class="emotion-2 emotion-3"
      type="checkbox"
    />
    <span>
      MyToggleSwitch
    </span>
  </label>
</div>
`;
